server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailSetMethod.scala

```diff
@@ -124,6 +159,77 @@ class EmailSetMethod @Inject()(serializer: EmailSetSerializer,
       .collectSeq()
       .map(DestroyResults)
 
+  private def update(emailSetRequest: EmailSetRequest, mailboxSession: MailboxSession): SMono[UpdateResults] = {
+    emailSetRequest.update
+      .filter(_.nonEmpty)
+      .map(update(_, mailboxSession))
+      .getOrElse(SMono.just(UpdateResults(Seq())))
+  }
+
+  private def update(updates: Map[UnparsedMessageId, EmailSetUpdate], session: MailboxSession): SMono[UpdateResults] = {
+    val validatedUpdates: List[Either[UpdateFailure, (MessageId, EmailSetUpdate)]] = updates
+      .map({
+        case (unparsedMessageId, updatePatch) => EmailSet.parse(messageIdFactory)(unparsedMessageId)
+          .map((_, updatePatch))
+          .toEither
+          .left.map(e => UpdateFailure(unparsedMessageId, e))
+      })
+      .toList
+    val failures: List[UpdateFailure] = validatedUpdates.flatMap({
+      case Left(e) => Some(e)
+      case _ => None
+    })
+    val validUpdates: List[(MessageId, EmailSetUpdate)] = validatedUpdates.flatMap({
+      case Right(pair) => Some(pair)
+      case _ => None
+    })
+
+    for {
+      updates <- SFlux.fromPublisher(messageIdManager.messagesMetadata(validUpdates.map(_._1).asJavaCollection, session))
+        .collectMultimap(metaData => metaData.getComposedMessageId.getMessageId)
+        .flatMap(metaData => {
+          SFlux.fromIterable(validUpdates)
+            .flatMap[UpdateResult]({
```

Comment:
Use concatMap here to not create contention on the modseq generation (monotic integer generation relying on optimistic concurrency)